tools/ast_gen.cpp

```diff
@@ -28,33 +30,92 @@ struct Node {
 
 int main() {
   std::vector<Node> nodes{
-      {"Binary", {{"Expr", "left"}, {"Token", "operatr"}, {"Expr", "right"}}},
-      {"Grouping", {{"Expr", "expression"}}},
-      {"Unary", {{"Token", "operatr"}, {"Expr", "right"}}}};
+      {"Binary",
+       {{"std::unique_ptr<Expr>", "lhs"},
+        {"Token", "op"},
+        {"std::unique_ptr<Expr>", "rhs"}}},
+      {"Grouping", {{"std::unique_ptr<Expr>", "expr"}}},
+      {"String", {{"std::string", "value"}}},
+      {"Number", {{"double", "value"}}},
+      {"Boolean", {{"bool", "value"}}},
+      {"Nil", {{"std::nullptr_t", "value"}}},
+      {"Unary", {{"Token", "op"}, {"std::unique_ptr<Expr>", "expr"}}}};
 
   fmt::println("// SPDX-License-Identifier: Apache-2.0");
   fmt::println("//");
   fmt::println("// SPDX-FileCopyrightText: Copyright (c) assignUser");
   fmt::println("");
   fmt::println("#pragma once");
-  fmt::println("#include \"lox/scanner.hpp\"");
+  fmt::println("#include <memory>");
+  fmt::println("#include <cstddef>");
+  fmt::println("");
+  fmt::println("#include \"lox/token.hpp\"");
   fmt::println("");
   // forward declare expressions for use in the abstract classes
-  std::ranges::for_each(nodes, [](Node& node){
-    fmt::println("class {};", node.type);
+  std::ranges::for_each(
+      nodes, [](Node &node) { fmt::println("class {};", node.type); });
+  fmt::println("");
+  fmt::println("class Visitor {{");
+  fmt::println("public:");
+  fmt::println("  virtual ~Visitor() = default;");
+  std::ranges::for_each(nodes, [](Node &node) {
+    fmt::println("  virtual void visit({0} const &expr) = 0;", node.type);
   });
+  fmt::println("}};");
   fmt::println("");
   fmt::println("class Expr {{");
   fmt::println("public:");
   fmt::println("  virtual ~Expr() = default;");
-  fmt::println("  virtual void accept( ExprVisitor& visitor ) const = 0;");
+  fmt::println("  virtual void accept(Visitor &visitor) const = 0;");
   fmt::println("}};");
   fmt::println("");
 
-  auto class_head = [&](std::string type) {
-    fmt::println("class {} : Expr {{", type);
+  auto class_head = [&](Node node) {
+    fmt::println("class {} : public Expr {{", node.type);
     fmt::println("public:");
+    // ctor
+    fmt::print("{}{}(", node.fields.size() > 1 ? "  " : "  explicit ", node.type);
+    std::string args =
+        std::accumulate(node.fields.begin(), node.fields.end(), std::string{""},
+                        [&](std::string acc, Field &field) {
+                          std::string comma = acc.size() == 0 ? "" : ", ";
+                          return std::move(acc) + comma +
+                                 fmt::format("{} {}", field.type, field.name);
```

Comment:
* prefer passing values by value or const-reference.
* prefer `std::string_view` for string literal constants
* we cannot pass acc by value since `std::accumulate` doesn't automatically move the lhs

Here's one possible correction:
```suggestion
                        [&](std::string& acc, Field const& field) {
                          if (not acc.empty()) {
                            acc.append(", ");
                          }
                          acc.append(field.type);
                          acc.append(' ');
                          acc.append(field.name);
                          return acc;
```